File: javaScript/linkedList/reverseLinkedList.js
/* 206. Reverse Linked List

Given the head of a singly linked list, reverse the list, and return the reversed list.

Example 1:
Input: head = [1,2,3,4,5]
Output: [5,4,3,2,1]

Example 2:
Input: head = [1,2]
Output: [2,1]

Example 3:
Input: head = []
Output: [] */

const reverseList = (head) => {
  let temp = head;
  let prev = null;
  while (temp !== null) {
    let next = temp.next;
    temp.next = prev;
    prev = temp;
    temp = next;
  }
  return prev;
};

console.log(reverseList([1, 2, 3, 4, 5])); //5,4,3,2,1

File: python/roadmap/singly & doubly linked lists/reverseLinkedList.py
""" 
Given the head of a singly linked list, reverse the list, and return the reversed list. 
Input: head = [1,2,3,4,5]
Output: [5,4,3,2,1]

"""

from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

    def __repr__(self):
        vals = []
        curr = self
        while curr:
            vals.append(str(curr.val))
            curr = curr.next
        return " -> ".join(vals)
        

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        
        curr = head
        prev = None

        while curr:
            temp_next = curr.next
            curr.next = prev
            prev = curr
            curr = temp_next

        return prev 
        

        

if __name__ == "__main__":
    head = ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5)))))
    print(Solution().reverseList(head))